Add router (connected-react-router)

ryanlaws opened the pull request:
This adds react-router and supporting packages and: 
- Integrates them into the root reducer and middleware. 
- Introduces a URL handler bound to state, which: 
  - Sets the URL according to the path of the last script loaded into the REPL
  - Attempts to load the script in the REPL on page load if the URL contains such a path.
- Adds an `async` action to recursively open the directory hierarchy containing the script. 
- Adds support utilities to convert between a file path and a URL path. 
- `fetch` for REPL endpoints now looks in both `/` and `/maiden` before failing.

ryanlaws commented:
Oops, I was testing with React's dev server and spaced on 404 handling which is going to require some modifications to the server. I'm a Go noob but I'm investigating. Looks like `gin` might need to be updated - https://github.com/gin-gonic/gin/pull/1663

ryanlaws commented:
Looks like `codec` also needs either an update or a replace. The former seems cleaner.
https://github.com/gin-gonic/gin/issues/1673

ryanlaws commented:
OK, I think that oughtta do it.

ryanlaws commented:
I missed `units.json` as well. My `fetchJsonFromRoot` implementation looks a little hacky - suggestions welcome.

ngwese commented on cmd/server.go:
```diff
@@ -92,6 +92,11 @@ func serverRun() {
 	})
 
+	// defer routing to web app
+	r.NoRoute(func(ctx *gin.Context) {
```
If I'm understanding the `NoRoute` directive it is basically overriding the default 404 response and instead returning the application index page. I'm concerned this will hamper debugging and potentially trigger some confusion elsewhere in code - for example if I have `http://norns.local/blah/balh/blah` the backend responds with the main index page (for the maiden app) to being loaded *but* with the bad URL as the location.

One pattern which I've seen in single page apps is that app side URL routing uses fragment syntax which makes it more clear which part of the stack is doing the routing; for example `http://norns.local/maiden#<stuff>`

ryanlaws replied:
Yes, exactly. I think the fragments are a much better idea, as long as they don't break back/forward in history.

ryanlaws replied:
Removed server/go changes.

ngwese commented on web/public/index.html:
```diff
@@ -5,9 +5,9 @@
-    <link rel="icon" type="image/png" href="maiden_96.png" sizes="96x96">  
+    <link rel="icon" type="image/png" href="/maiden/maiden_16.png" sizes="16x16">  
+    <link rel="icon" type="image/png" href="/maiden/maiden_32.png" sizes="32x32">  
+    <link rel="icon" type="image/png" href="/maiden/maiden_96.png" sizes="96x96">  
```
If we used fragments in the URLs I think we could bypass the this change and retain the ability to host maiden from an arbitrary root

ryanlaws replied:
Yeah, I think the fragments idea solves a lot of that and would eliminate the need for a lot of these auxiliary changes.

ryanlaws replied:
Removed all changes here.

ngwese commented on web/src/model/edit-actions.js:
```diff
@@ -213,9 +213,32 @@ export const bufferRead = resource => dispatch => {
+// This is async because nested directories don't 
+// like opening until their parents have loaded
+export const directoryReadRecursive = resource => async dispatch => {
+  await resource.split('/').reduce(async (asyncUrl, asyncPart) => {
```
Using the test/dev setup documented in the readme I was getting stack traces from React. If the starting URL was `http://localhost:3000` (the default for `yarn start`) then the `resource` value was undefined so the call to `split(...)` raised an exception. If starting from `http://localhost:3000/maiden` then things were okay.

I experimented with just wrapping the whole await expression inside `if (resource) { ... }` and that fixed it but that felt a tiny bit hacky to me... not sure what your thoughts are there.

ryanlaws replied:
Ack. Yeah, I started with the `yarn start` dev server. When I tried serving straight from norns, the paths didn't match (`/` vs. `/maiden`, hence the try-catch), and that's when I started messing with the 404 code, but I didn't go back and try the dev server again. 

I don't think checking whether `resource` is truthy is too hacky, for my tastes at least. I'd prefer bailing if `!resource` over a big `if` block to reduce indentation, but that's well above the threshold of nitpickiness and I'd defer your preference on that.

ryanlaws replied:
Did the check for `!resource`.

ngwese commented on web/src/utils.js:
```diff
@@ -7,3 +7,22 @@ OS.metaKey = function metaKey() {
+export const pathToResource = (path) => {
+  const sub = findSubPath(path);
+  return sub ? `${apiBase}${dustBase}${sub}` : null;
+}
```
These constants partly duplicate values which are already defined in the `api.js` file. It would be preferable to use those:
https://github.com/monome/maiden/blob/aff75f1243451948b0e2053fb13c22dfb77d2618/web/src/api.js#L3-L13

It would be worth using (and refactoring as needed) the various static functions in api.js which are doing similar operations; for example functions like `resourceForScript(...)` and `fileFromResource(...)` seem similar-ish.

ryanlaws replied:
Oh, nice, that's really helpful. This is my first time poking around in here so I'm not surprised I glossed over existing constants.

ryanlaws replied:
I exported the `API_ROOT` constant here. I'm not totally seeing the forest for the trees with the refactoring of those functions yet. I might need some more concrete guidance there if more changes are needed.

ryanlaws replied:
OK I think I see what you meant now. I didn't refactor or add anything to `api.js` but I now use the static functions you mentioned. I considered moving some of the `url-utils.js` functions to `api.js`, but I thought it didn't fit their responsibility of translating between fragments and resources. I'm not 100% certain what all is in the purview of `api.js` but I didn't want to pollute it.

I added `url-utils.test.js` to assist with the refactor and I also went ahead and added a test indicated as missing by a `TODO` in `api.test.js`. I also took the liberty of wrapping the existing tests in a `describe`. I hope I'm not overstepping any boundaries here.

ryanlaws commented:
Great feedback! Thank you for taking a look at this! I probably won't be able to dig in right away, but I think using a fragment instead makes a lot of sense as long as it persists to history. If we can avoid messing with the server code, that certainly seems preferable - that definitely felt a little wrong.

ryanlaws commented:
> * in the future we could add /<n> after the resource to allow multiple editor views...?

Hmm, I'm not completely following here. Is <n> just an integer? Is it currently possible to split the editor view or something?

> One other thing I tried which didn't work and was possibly confusing was that while browser history seemed to be enabled using the back button in the browser window would show the URL of the previous file I had directly edited but neither the editor itself nor the file tree changed to display the previous file.

Good call. That was super funky and I completely missed it in my testing. I trashed the `componentDidMount` for page load and made it use state instead and it seems much cleaner to me now. 

Let me know if you think this needs more changes. This feels pretty good to me now, but it's quite possible I missed something else. A couple of things you mentioned sound intriguing but weren't completely clear to me. They might make more sense in the morning.

ryanlaws commented:
@ngwese I refactored and added some tests and resolved the above conversations. Let me know whether this fits what you had in mind, or still needs work.

ngwese commented:
I will take a deeper at the code in a bit but one of the things which I noticed quickly was that something funky is going on with the refresh of the file tree now (I don't think this was in the previous version). It appears as if went switching files where is some code which runs async and tries to refresh the tree which causes the previous expanded portion of the tree to collapse/disappear but the disclosure triangles still indicate it is expanded.

_I captured a video but apparently I can't attach it the PR so..._
http://afofo.com/files/norns/maiden-router-file-tree-glitch.mov

ryanlaws commented:
Ah, great, thanks for the video, that is super helpful! Earlier I noticed some similar funkiness when loading scripts in the same directory, which is what the `isSameDir` and its consumer are meant to address. It makes sense that you'd see that in a sibling directory. That whole expanding thing probably only makes sense to do on page load (from fragment). 

I was able to repro the behavior from your video and I've reworked it and as a bonus managed to factor out and remove `isSameDir`.

ngwese commented:
It all looks good and working on my end!

I squashed down a few of the intermediate commits where the router was still using `/` instead of `#` in order to avoid some confusion but left the rest of the commit history alone. I did the partial squash and merge outside of GH in such a way that it appears GH no longer associates it with this PR.

Closing out this PR now that the changes have been merged.

ryanlaws commented:
Cool, thank you!
